### ISSUE TITLE
draft: discover and identify a vulnerabilty

### ISSUE THREAD

**Issue by aleeusgr** [haskell/security-advisories#134]
Hello!
I am solving Advent of Code and I came across an effect I never encountered before in Haskell: my **code runs, but it makes the computer hang**. It's a bit hard to reproduce, since you need to get personalized inputs file from AoC, but I have one I can share. 

- [ ] provide list of inputs
- [ ] document the program
- [ ] ID the package and the version I am using.
- [ ] ID cvss

The changes that I made debugging the function at the line 42, on my list of inputs it reports an error that I am trying to call `head` on an empty list (or a singleton list?)

Code example is in the branch but again I just realized it won't run without the inputs.
https://github.com/aleeusgr/aoc23/tree/haskell-avisory-report 

Mandatory information:

* Package : `?`
* cvss: `?`
* affected versions: `?`

Optional:

* cve: `?`
* keywords: `?`
* aliases: `?`
* related: `?`
* affected OSes: `?`
* affected architecture: `?`
* declarations: `?`
* Long description: `?`

**Comment by frasertweedale** [comment]
@aleeusgr tip of `main` branch of your repo does not build:
```
src/MyLib.hs:42:30: error:
    • Couldn't match expected type ‘[Char]’ with actual type ‘Char’
    • In the expression: x
      In an equation for ‘findAndReplaceNumber’:
          findAndReplaceNumber [x] _ = x
   |
42 | findAndReplaceNumber [x] _ = x
   |                     
```

Could you please let me know the commit to build, and attach an input that triggers the condition.

I wonder if it would be a space leak that eats all your system memory and causes issues?  Anyway, I would be happy to help identify the issue.

**Comment by aleeusgr** [comment]
Thanks!
It will take me some time, it's the first opportunity I see to use property testing to find the bug. I want to learn QuickCheck and I want to find the bug probably the same way that you did. Thanks for the inspiration 🙏

**Comment by aleeusgr** [comment]
This issue traces back to [ghc](https://gitlab.haskell.org/ghc/ghc/-/issues/9207) and seems to be not a security issue.

<details><summary>Details</summary>
<p>

I used ghci debugger to realize that one of the lower level functions breaks on empty string. 
Which pointed me to the buggy implementation: 
```haskell
wordContainsDigits word = or $ map (`isSubsequenceOf` word) ["one", ...]
```
So I used QuickCheck to find small inputs that break the high level function:
```haskell
QC.testProperty "correctCalibrationValues returns ..." $
\x -> length (MyLib.correctCalibrationValues x :: [Char] )  <= length x
```
Turns out:
```haskell
ghci> "two" `isSubsequenceOf` "twao"
True
```

This allowed incorrect inputs to be processed by the function, which ran out of items to check and was looping infinitely by adding an empty string to the incorrect input, which passed the check and higher level and got passed down again to the parser.


</p>
</details>
